build: add some gcc flags to clang

dyfer opened the pull request:
<!-- For information about contributing see: https://github.com/supercollider/supercollider/wiki/Contributing-directory -->

## Purpose and Motivation

<!-- Please provide a description, even if you are linking to a related Issue or PR. -->
<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->

This adds flags that I _think_ are valid for both compilers to the clang section.

I don't know how to enable something like `-Werror=return-type` on MSVC... or if any of the other flags have equivalents.

## Types of changes

<!-- Delete lines that don't apply -->

- New feature (?)

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [ ] Code is tested
- [ ] All tests are passing
- [ ] Updated documentation
- [x] This PR is ready for review <!-- If not ready for review, consider making a Draft PR instead, or describe what kind of specific feedback you want. -->

<!-- ## Merge notes
(optional) You may also add notes of:
     - Dependencies: if this PR depends on any other PRs or actions
     - Merge Note: any considerations regarding merging this PR
     - Release Notes: text you think would be useful to include in the Release Notes
-->

<!-- Note: The recommended commit message format is:
     component: sub-component: short description of changes
For example:
     classlib: ClassBrowser: fix search with empty query string
Common `comp:` labels:
     help, classlib, sclang, scsynth, plugins, tests, docs, SCDoc
See wiki for more info: Wiki->Creating Pull Requests->git-commit-message-format
-->

dyfer committed:
Update nova-simd

dyfer committed:
build: add some gcc flags to clang

JordanHendersonMusic commented on CMakeLists.txt:
```diff
@@ -340,11 +340,8 @@ endif()
 if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
 	add_compile_options(
-		-Werror=return-type      # Error out if function definitions lack a return type
 		-fno-math-errno          # Optimize math functions by not setting errno
 		-fno-signaling-nans      # Assume no signaling NaNs for better optimization
```
I think we want both of these on clang too.

JordanHendersonMusic replied:
But no-signaling-nans is called 

no-signaling-math[¶](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fsignaling-math)

dyfer replied:
no-math-errno was causing compilation issues, like:

```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.1.sdk/usr/include/c++/v1/cmath:322:5: error:

      <cmath> tried including <math.h> but didn't find libc++'s <math.h> header. This usually means that your header search paths
      are not configured properly. The header search paths should contain the C++ Standard Library headers before any C Standard
      Library, and you are probably using compiler flags that make that not be the case.
  322 | #   error <cmath> tried including <math.h> but didn't find libc++'s <math.h> header. \
      |     ^

fatal error: too many errors emitted, stopping now [-ferror-limit=]
```

Any thoughts on this?

And should I add `no-signaling-math` to clang only?

JordanHendersonMusic replied:
Ah okay this is a little complicated..

clang does not have `-fno-signaling-nans`

ignore what I said about `no-signaling-math` that is wrong, don't use that.

Perhaps just set `-fno-signaling-nans` on GCC. 

Re `no-math-errno `

> On some targets, math library functions never set errno, and so -fno-math-errno is the default. This includes most BSD-derived systems, including Darwin.

https://clang.llvm.org/docs/UsersManual.html

I think this means we want to set it, unless we are on apple where it seems to cause an error!?

dyfer replied:
I added it to clang with a conditional switch disabling it on apple.

JordanHendersonMusic commented on CMakeLists.txt:
```diff
@@ -363,6 +360,12 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
 if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
+	add_compile_options(
+		-Werror=return-type      # Error out if function definitions lack a return type
+		-fsigned-zeros           # Maintain distinction between -0.0 and +0.0
+		-fno-associative-math    # Prevent reordering of floating-point operations
```
We should also specify mscv too.
I ***believe*** the correct flag would be /fp:precise 

https://learn.microsoft.com/en-us/cpp/build/reference/fp-specify-floating-point-behavior?view=msvc-170

dyfer replied:
/fp:precise seems to be the default, should we still set it?

JordanHendersonMusic replied:
I think there should at least be a comment here saying that is what we support and that if we ever add something, we need to make sure msvc does the same thing.

dyfer replied:
I added the comment

JordanHendersonMusic reviewed (changes requested):
I think this looks good, just to check though that the submodule update is supposed to go in this PR?

dyfer commented:
> just to check though that the submodule update is supposed to go in this PR?

Yes! That pulls in the [change in nova-simd which fixed building on clang/arm64](https://github.com/timblechmann/nova-simd/pull/12) with `-Werror=return-type`. The submodule also had a couple of cosmetic changes since the last time we updated it, that's why the diff is a little larger.

JordanHendersonMusic reviewed (approved):
okay looks good

dyfer commented:
I also noticed that we set [`-fno-finite-math-only`](https://github.com/supercollider/supercollider/blob/develop/server/plugins/CMakeLists.txt#L61) for the plugins for GCC.
- should we touch it in this PR and apply it to Clang as well?
- if so, should that flag stay for plugins only or be used throughout?

JordanHendersonMusic commented:
Actually I think it should be removed because it is the default. 


>-ffinite-math-only
Allow optimizations for floating-point arithmetic that assume that arguments and results are not NaNs or +-Infs.
This option is not turned on by any -O option besides -Ofast since it can result in incorrect output for programs that depend on an exact implementation of IEEE or ISO rules/specifications for math functions. It may, however, yield faster code for programs that do not require the guarantees of these specifications.
The default is -fno-finite-math-only.

https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html

I think it is pretty safe to just say we do not support -Ofast, only -O3, with the few tweaks we have already added.


Seems it was originally added all the way back in 2010 https://github.com/supercollider/supercollider/commit/018a3b5a69b428c7ff8c0888737d067250abe59b 

so the gcc defaults might have changed since then which might explain why it was added in the first place.

dyfer commented:
> Seems it was originally added all the way back in 2010 [018a3b5](https://github.com/supercollider/supercollider/commit/018a3b5a69b428c7ff8c0888737d067250abe59b)
> 
> so the gcc defaults might have changed since then which might explain why it was added in the first place.

Ah, I see. In that case it might've been added because [we used to compile with -ffast-math](https://github.com/supercollider/supercollider/commit/018a3b5a69b428c7ff8c0888737d067250abe59b#diff-34dc2e44b14fc26c8651b204eaf2195382705e6526c6143edaa60a2e464c5e45R5), which was eventually removed in https://github.com/supercollider/supercollider/pull/4119.

So should I remove -fno-finite-math-only flag in this PR?

JordanHendersonMusic commented:
Ah that makes more sense! 

Yes I think since you are touching the maths stuff, it should go here.

dyfer commented:
I updated the PR. Still good to merge? (once CI passes)

dyfer commented:
@JordanHendersonMusic just to double check, can I merge this after also removing flags from [server/plugins/CMakeLists.txt](https://github.com/supercollider/supercollider/pull/7263/files#diff-3d1ccc9e1c72318a4384e2c1430e41a1b02019a0352cff7dbe4e6791eac2f9b7)?

JordanHendersonMusic commented:
Yes I think that will be fine since it was never enabled - merge away!
